docs/support/tutorials/nextflow-puhti.md

```diff
@@ -1,191 +1,309 @@
+         --cores 4 \
+         --output results \
+         --profile local,singularity \
+         --cachedir ${SCRATCH}/singularity \
+         --databases ${SCRATCH}/databases/WtP_databases \
+         --workdir ${SCRATCH}/workflow-phages-username
+```
+
+Please note that the container images in this example should be in the folder
+`/scratch/<project>/What_the_Phage/singularity`. Otherwise, Nextflow tries to
+download the images again which increases the likelihood of failures. Remember
+to edit `<project>` to match your actual CSC computing project name.
+
+## 7. Demonstration of nf-core Nextflow pipeline using HyperQueue executor (optional)
+
+In this example, let's use the
+[HyperQueue meta-scheduler](../../apps/hyperqueue.md) for executing a Nextflow
+pipeline. This executor can be used to scale up analysis across multiple nodes
+when needed.
+
+Here is a batch script for running a nf-core pipeline on Puhti:
+
+```bash
+#!/bin/bash
+#SBATCH --job-name=nextflowjob
+#SBATCH --account=<project>
+#SBATCH --partition=small
+#SBATCH --time=00:10:00
+#SBATCH --nodes=1
+#SBATCH --ntasks-per-node=1
+#SBATCH --cpus-per-task=40
+#SBATCH --mem-per-cpu=2G
+  
+module load hyperqueue/0.16.0
+module load nextflow/22.10.1
+
+# create a directory on scratch folder for running nextflow pipeline
+export SCRATCH=/scratch/<project>/$USER/nextflow
+mkdir -p ${SCRATCH}
+cd ${SCRATCH}
+
+export APPTAINER_TMPDIR="${SCRATCH}"
+export APPTAINER_CACHEDIR="${SCRATCH}"
+unset XDG_RUNTIME_DIR
+
+# Specify a location for the HyperQueue server
+export HQ_SERVER_DIR=${PWD}/hq-server-${SLURM_JOB_ID}
+mkdir -p "${HQ_SERVER_DIR}"
+
+# Start the server in the background (&) and wait until it has started
+hq server start &
+until hq job list &>/dev/null ; do sleep 1 ; done
+
+# Start the workers in the background and wait for them to start
+srun --overlap --cpu-bind=none --mpi=none hq worker start --cpus=${SLURM_CPUS_PER_TASK} &
+hq worker wait "${SLURM_NTASKS}"
+
+# As an example, let's clone a nf-core pipeline and run a test sample
+git clone https://github.com/nf-core/rnaseq.git -b 3.10
+cd rnaseq
+nextflow run main.nf -profile test,singularity --outdir . -resume
+
+# Wait for all jobs to finish, then shut down the workers and server
+hq job wait all
+hq worker stop all
+hq server stop
 ```
 
-Please note that the Singularity images in this example should be in the folder:
-`"/scratch/project_xxx/What_the_Phage/singularity"`. Otherwise, Nextflow tries to download 
-the images again which increases the likelihood of failures. Remember to edit `project_xxx`
-to match your computing project.
+!!! note
+     In case you are using multiple nodes, ensure that you use the right
+     executor and that the executor knows how many jobs it can submit. For
+     this, below is an example snippet that one should add to the
+     `nextflow.config` file:
+
+     ```text
+     executor {
+     queueSize = 40*SLURM_NNODES
+     name = 'hq'
+     cpus = 40*SLURM_NNODES
+     }
+     ```
```

Comment:
Would it be best to add this to the batch script similar to how it is done in the other HyperQueue example? As far as I understand, without it the above batch script won't actually use HyperQueue as executor by default?

Reply:
I added a suggestion related to this. Regardless of how we do it, I think the steps needed should be made more explicit somehow, since the batch script does not automatically use HQ (and it also seems to timeout). I think it may be useful for someone who already understands a bit how Nextflow works and is able to edit the batch script, but perhaps not a complete beginner?

Reply:
Yes. I agree.